apps/ex_wire/lib/ex_wire/p2p/manager.ex

```diff
@@ -133,64 +140,74 @@ defmodule ExWire.P2P do
             "[Network] [#{peer}] Failed to read incoming packet from #{peer.host} `#{reason}`)"
           )
 
-        conn
+        %{conn | last_error: reason}
     end
   end
 
-  defp handle_packet(packet, session, conn) do
-    if DEVp2p.session_active?(session) do
-      do_handle_packet(packet, conn)
-      session
-    else
-      attempt_session_activation(session, packet)
+  @spec handle_packet(module(), Packet.packet(), Connection.t()) :: Connection.t()
+  defp handle_packet(packet_mod, packet, conn) do
+    packet_handle_response = packet_mod.handle(packet)
+
+    if not DEVp2p.session_active?(conn.session) && packet_handle_response != :activate do
+      Logger.error(fn -> "Expected message prior to activation, got: #{inspect(packet)}" end)
     end
-  end
 
-  defp do_handle_packet(packet = %Packet.Status{}, conn) do
-    return_packet =
-      case Packet.Status.handle(packet) do
-        :ok ->
-          Packet.Status.new(packet)
+    case packet_handle_response do
+      :ok ->
+        conn
 
-        {:disconnect, reason} ->
-          Packet.Disconnect.new(reason)
-      end
+      :activate ->
+        new_session =
+          if DEVp2p.session_active?(conn.session) do
```

Comment:
maybe pulling logging out of the function makes more sense.

Reply:
Good point, I'll try and reduce side-effects here and leave that to the caller maybe?